server/search/search.go

```diff
@@ -0,0 +1,285 @@
+	indexMapping := bleve.NewIndexMapping()
+	for _, config := range configs {
+		indexMapping.AddDocumentMapping(config.Name, newDocumentMapping(config))
+	}
+	// Disable default mapping as mapping is done manually.
+	indexMapping.DefaultMapping = bleve.NewDocumentDisabledMapping()
+	indexMapping.StoreDynamic = false
+	indexMapping.DocValuesDynamic = false
+	indexMapping.IndexDynamic = false
+	return indexMapping
+}
+
+// IndexEntities documents from given quad store by given configuration to given index
+// If entities is empty index all the matching documents in the quad store
+// Otherwise only load documents of the provided entities
+func IndexEntities(ctx context.Context, qs graph.QuadStore, configs Configuration, index Index, entities ...quad.Value) error {
+	batch := index.NewBatch()
+	for _, config := range configs {
+		clog.Infof("Retreiving documents for \"%s\"...", config.Name)
+		documents, err := getDocuments(ctx, qs, config, entities)
+		if err != nil {
+			return err
+		}
+		for _, document := range documents {
+			id := document.id
+			batch.Index(id, document.data)
+		}
+		clog.Infof("Retrieved %v documents for \"%s\"", len(documents), config.Name)
+	}
+	clog.Infof("Indexing %v documents", batch.Size())
+	index.Batch(batch)
+	return nil
+}
+
+// Delete documents from given quad store by given configuration for given index
+// If entities is empty delete all matching documents in the quad store
+// Otherwise only delete documents of the provided entities
+func Delete(ctx context.Context, qs graph.QuadStore, configs Configuration, index Index, entities ...quad.Value) error {
+	batch := index.NewBatch()
+	for _, config := range configs {
+		clog.Infof("Retreiving documents for \"%s\"...", config.Name)
+		documents, err := getDocuments(ctx, qs, config, nil)
+		if err != nil {
+			return err
+		}
+		for _, document := range documents {
+			id := document.id
+			batch.Delete(id)
+		}
+		clog.Infof("Retrieved %v documents for \"%s\"", len(documents), config.Name)
+	}
+	clog.Infof("Deleting %v documents", batch.Size())
+	index.Batch(batch)
+	return nil
+}
+
+// NewIndex builds a new search index
+func NewIndex(ctx context.Context, qs graph.QuadStore, configs Configuration) (Index, error) {
+	clog.Infof("Building search index...")
+	indexMapping := newIndexMapping(configs)
+	index, err := bleve.New(IndexPath, indexMapping)
+	if err != nil {
+		return nil, err
+	}
+	err = IndexEntities(ctx, qs, configs, index)
+	if err != nil {
+		return nil, err
+	}
+	clog.Infof("Built search index")
+	return index, nil
+}
+
+// GetIndex attempts to open an existing index, if it doesn't exist it creates a new one
+func GetIndex(ctx context.Context, qs graph.QuadStore, configs Configuration) (Index, error) {
+	index, err := OpenIndex()
+	if err == bleve.ErrorIndexPathDoesNotExist {
+		return NewIndex(ctx, qs, configs)
+	}
+	return index, err
+}
+
+// ClearIndex removes the existing index
+func ClearIndex() error {
+	err := os.RemoveAll(IndexPath)
```

Comment:
Should be inside `kv` package. Same for almost everthing else in this package except `Path`s. Those should be in `query/path` package and use the `shape.Search` added in this PR.